Import `generalized-bulletproofs`

kayabaNerve opened the pull request:
This is part of the `fcmp++` branch which has been audited, and as a first step to merging the `fcmp++` branch, is being imported by itself. The commits explain how to verify the process of the migration _and_ the relation to the audited codebase.

Adds a MSRV bound of <= 1.69 for generalized-bulletproofs via a GitHub action, but does not enforce it via the MSRV policy. It's just to let us know if the MSRV exceeds 1.69, at which point we can respond accordingly. That may cause this soft bound to affect as if a hard bound though?

Draft as I still have tweaks to make before it's ready to merge. Please _merge commit_ when done to preserve the commit hashes of this PR (as they're referenced within the documentation).

kayabaNerve committed:
Import `generalized-bulletproofs`

`git checkout 54da48f27a05fa8656014942919da1dfbab4d8e3 ./crypto/generalized-bulletproofs`
replicates this commit in its entirety. It solely copies/pastes the
`generalized-bulletproofs` library present on the current (as of the time of
writing) HEAD of the fcmp++ branch of
https://github.com/monero-oxide/monero-oxide.

`generalized-bulletproofs` was originally developed under
https://kayabaNerve/fcmp-plus-plus, the original repository for Monero's FCMP++
upgrade's implementation. The library itself was audited by Aaron Feickert, who
also wrote the security proofs for Generalized Bulletproofs. The security
proofs can be found here:
https://github.com/cypherstack/generalized-bulletproofs
The audit can be found here:
https://github.com/cypherstack/generalized-bulletproofs-code

The next commit will add those to the `audits` directory.

The audit was performed as of commit 7d972e2b16fe57f4dfa4a9549d2fbb9306411704
of the FCMP++ repository, as specified within the audit. The code present here
can be compared as follows.

```
git clone https://github.com/kayabaNerve/fcmp-plus-plus # Now archived
cd fcmp-plus-plus
git checkout 7d972e2b16fe57f4dfa4a9549d2fbb9306411704
cd ..
git diff --no-index ./fcmp-plus-plus/crypto/generalized-bulletproofs ./crypto/generalized-bulletproofs
```

The most notable change is the replacement of `reduce_512` with
`FromUniformBytes<64>`, and the optimization of the prover (which did NOT
modify the verifier).

kayabaNerve committed:
Import `generalized-bulletproofs` audits

`git checkout -f 54da48f27a05fa8656014942919da1dfbab4d8e3 ./audits/generalized-bulletproofs`

Same premise as the prior commit.

kayabaNerve committed:
Add generalized-bulletproofs to the workspace

kayabaNerve committed:
Update `generalized-bulletproofs` LICENSE

kayabaNerve committed:
Update generalized-bulletproofs README

kayabaNerve commented:
clippy failing due to a lint disabled in #31

kayabaNerve committed:
Fix relative links with spaces in their names

kayabaNerve committed:
Add a bound in the workflow that generalized-bulletproofs compiles with Rust 1.69

This effectively changes our MSRV policy by somewhat committing to a Rust 1.69
target for this library. If in the future, we have a PR which causes this
action to fail, we'd have to decide whether we want to enforce it as a
requirement or remove the bound.

Simultaneously performs the `cargo update`s required to achieve this MSRV, and
updates the `generalized-bulletproofs` Cargo.toml to require the necessary
dependency versions. It already specified the same minor versions, but these
patch versions are necessary for the MSRV to be satisfied.

kayabaNerve committed:
Remove potential panics in `generalized-bulletproofs/src/transcript.rs`

kayabaNerve committed:
Remove `ScalarVector` from the public API

The provided methods panicked on any behavior considered invalid. It's trivial
to remove from the public API, resolving all panic concerns, even though it
reduces the functionality of the library.

kayabaNerve committed:
Don't export PointVector

It was declared public, and exported, but was never actually used in the public
API.

Uncomments the various commented functions present here without real purpose.
The other option would be to delete them, but they're decent, simple utilities
which are well implemented.

kayabaNerve committed:
Update documentation in lincomb, de-duplicating a brief segment of code

Also removes the NOP Zeroize from Variable. Calls should use `#[zeroize(skip)]`
for Variable.

kayabaNerve committed:
Move the arithmetic circuit test to be external, off the public API

kayabaNerve committed:
Rename generators to insecure_test_generators, remove panics

kayabaNerve committed:
Remove potential panics in `crypto/generalized_bulletproofs/src/lib.rs`

Potential overflows and a panic if a specified generator was the point at
identity.

kayabaNerve committed:
Remove potential panics in crypto/generalized-bulletproofs/src/inner_product.rs

kayabaNerve committed:
Split IpError/AcErrror into IpProveError, IpVerifyError, AcProveError, AcVerifyError

kayabaNerve committed:
Add check inner-product challenges are non-zero

Statistically negligible probability, but the proof would be invalid if so.

kayabaNerve committed:
Clarify documentation and make slight optimizations within crypto/generalized-bulletproofs/src/arithmetic_circuit_proof.rs

kayabaNerve committed:
Fix no-std builds of generalized-bulletproofs

kayabaNerve commented:
clippy/msrv should be resolved by merging #31.

Unfortunately, our `Cargo.lock` is for some reason polluted with irrelevant elliptic curves thanks to the most recent update to `ciphersuite` (despite not enabling their optional features, as they were before?). A new (on the Serai tree) version of `ciphersuite` splits out the curves' implementations, leaving solely the trait, solving this _but_ it has yet to go through a release cycle.

I did make changes while ensuring this wouldn't panic. Mainly layout changes/updating the doc strings. Unfortunately, this wasn't a crate I polished before it was sent off for audits (as the audits were externally organized), so it had a bit more polish queued than most crates would.

kayabaNerve commented:
Merged main to incorporate #31 which should fix _this_ PR's CI.

kayabaNerve committed:
Add "crypto" branch as reason to run tests

kayabaNerve committed:
Support proving proofs with empty `aL`, `aR` again within `generalized-bulletproofs`

Regression introduced when padding was removed. 'empty' proofs are mapped to
`n = 1` proofs (as the smallest set of generators allowed is for `n = 1`, and
a `n = 0` proof is undefined).

These proofs technically have use for anyone who wants to prove a series of
commitments/constrain them, but never needs to calculate any products within
the witness.

kayabaNerve committed:
Remove `Debug` from the `generalized-bulletproofs` witness elements

kayabaNerve committed:
Scope `audits/generalized-bulletproofs` to `audits/crypto/generalized-bulletproofs`

kayabaNerve committed:
`doc_auto_cfg` -> `doc_cfg`

kayabaNerve committed:
Use `is_power_of_two` in applicable spot

Boog900 reviewed (commented):
Just some small nits, otherwise all looks ok

kayabaNerve committed:
Merge two `for` loops in `generalized-bulletproofs`, optimizing out a large allocation used to carry data between them

kayabaNerve commented on crypto/generalized-bulletproofs/src/arithmetic_circuit_proof.rs:
```diff
@@ -0,0 +1,765 @@
+      // Lines 85-87 calculate `WL`, `WR`, `WO`
+      // We preserve them in terms of `g_bold` and `h_bold` for a more efficient multiexp
+      let mut h_bold_scalars = l_weights * x[jlr];
+      for (i, wr) in (r_weights * x[jlr]).0.into_iter().enumerate() {
```
`jlr = ilr`, as `js = is`, hence their prior omission. It'd presumably be better to explicitly declare them as expected.

kayabaNerve committed:
Respond to a pair of nits raised by @Boog900

kayabaNerve committed:
Represent `WCG` not as a non-sparse `Vec` of sparse vectors, but a sparse `BTreeMap` of sparse vectors (#4)

* Represent `WCG` not as a non-sparse `Vec` of sparse vectors, but a sparse `BTreeMap` of sparse vectors

Within the Monero protocol, the non-sparse `Vec` was spending hundreds of MB on
empty vectors (just 16 bytes each) for commitments which had no terms
constrained.

An earlier attempt tried a truly sparse vector, yet that invoked a O(n^2)
algorithm to compute as the vector had to be iterated to find all entries
within a specific vector, causing an O(n) invocation per commitment. This
tripled the verification time.

This algorithm represents the memory in O(n log c) where n is the
constrained terms and c is the amount of commitments which have constrained
terms. While the amount of commitments is large, the amount constrained within
a single `LinComb` should be small. Processing is an `O(log c)` lookup before
the traditional `O(n)` iteration. Accordingly, this should have almost all of
the memory savings with almost none of the performance cost.

cc @j-berman to provide exact performance numbers for the 1-in, 2-in, 4-in,
8-in, and 128-in cases (verification and time to verify) on
https://github.com/monero-oxide/monero-oxide/pull/39. cc @boog900 to review
this change to ensure it doesn't change the wires of the encoded program.

* Adopt the `BTreeMap::entry` API as suggested by @boog900
